core/src/main/java/com/onelogin/saml2/util/Util.java

```diff
@@ -967,19 +981,35 @@ public static Boolean validateSignNode(Node signNode, X509Certificate cert, Stri
 			Element sigElement = (Element) signNode;
 			XMLSignature signature = new XMLSignature(sigElement, "", true);
 
+			String extractedFingerprint;
+			X509Certificate extractedCert = extractCertificate(signature);
+			if (extractedCert != null) {
+				extractedFingerprint = calculateX509Fingerprint(extractedCert, alg);
+			} else {
+				extractedFingerprint = null;
+			}
+
 			if (cert != null) {
-				res = signature.checkSignatureValue(cert);
+
+				// Issue #149, don't check signature if the certificate doesn't match the certificate in the XML
+
+				if (extractedFingerprint != null && !calculateX509Fingerprint(cert, alg).equals(extractedFingerprint)) {
+					// certificate is present in the signature, it must match the certificate that we have.
+					LOGGER.warn("Certificate used in the document does not match one of configured certificates");
```

Comment:
When we have multiple certificates, fake warnings could be throwed.

Reply:
Would you rather all warnings are printed in a batch after all of them fail? This will require dragging richer return type from Util.validateSignature()... In which case, I'd do it differently, and be more precise to the cause of error when throwing.

Reply:
No, it would be better throw warning only when all certify checking was failed.

Reply:
I'm sorry, can you please clarify?
`Util.validateSignature()` doesn't know that there are more certificates available. If error that explains its reason, is printed only when all certificates are checked, either all certificates need to be passed to `Util.validateSignature()`, or `Util.validateSignature()` needs to return more information to the caller, so the caller can collate the errors. Or am I missing something here?